templates/common/run/diff.go

```diff
@@ -35,101 +37,128 @@ import (
-		// return empty string.
-		file1Exists, err := common.Exists(file1)
-		if err != nil {
-			return "", err //nolint:wrapcheck
-		}
-		file2Exists, err := common.Exists(file2)
-		if err != nil {
-			return "", err //nolint:wrapcheck
-		}
-		if !file1Exists && !file2Exists {
-			return "", nil
-		}
+
+	switch exitCode {
+	case 0:
+		return "", nil
+	case 1:
+		return trimMetadata(stdout.String()), nil
+	default:
 		return "", fmt.Errorf("error exec'ing diff: %s", stderr.String())
 	}
-	return stdout.String(), nil
 }
 
-var (
-	diffColorOnce    sync.Once
-	diffColorSupport bool
-	diffColorErr     error //nolint:errname
-)
+// Copies absPath (if it exists) into tempDir at the given relative path. If
+// absPath doesn't exist, then an empty file is create at the given relative
+// path instead.
+func copyToTempIfExists(ctx context.Context, absPath, tempDir, linkRelPath string) error {
+	srcExists, err := common.Exists(absPath)
+	if err != nil {
+		return err //nolint:wrapcheck
+	}
 
-// diffColorSupported returns whether we're running on a machine that supports
-// --color=always. MacOS <= 12 seems not to have this.
-func diffColorSupported(ctx context.Context) (bool, error) {
-	diffColorOnce.Do(func() {
-		diffColorSupport, diffColorErr = diffColorCheck(ctx)
-	})
-	return diffColorSupport, diffColorErr
-}
+	dest := filepath.Join(tempDir, linkRelPath)
+	if err := os.MkdirAll(filepath.Dir(dest), common.OwnerRWXPerms); err != nil {
+		return err //nolint:wrapcheck
+	}
 
-// diffColorCheck tests whether we're running on a machine that supports
-// --color=always. MacOS <= 12 seems not to have this.
-func diffColorCheck(ctx context.Context) (bool, error) {
-	var stderr bytes.Buffer
-	opts := []*Option{
-		WithStderr(&stderr),
-		AllowNonzeroExit(),
+	if !srcExists {
+		return os.WriteFile(dest, nil, common.OwnerRWPerms) //nolint:wrapcheck
 	}
-	exitCode, err := Run(ctx, opts, "diff", "--color=always", "/dev/null", "/dev/null")
-	if err != nil {
-		return false, fmt.Errorf("failed determining whether the diff command supports color: %w", err)
+
+	if err := common.Copy(ctx, &common.RealFS{}, absPath, dest); err != nil {
+		return err //nolint:wrapcheck
 	}
+	return nil
+}
 
-	if exitCode == 2 && strings.Contains(stderr.String(), "unrecognized option `--color=always'") {
-		return false, nil
+// The git diff command includes extra metadata header lines that we don't want,
+// like this:
+//
+//	diff --git a/file1.txt b/file2.txt
+//	index 84d55c5..e69de29 100644
+//
+// ... so we remove them, but only if they are present at the beginning of the
+// file.
+//
+// These metadata lines aren't presented in regular non-diff git output, and
+// they're just clutter. Also, the "index" line leaks metadata that we don't
+// want to leak.
+func trimMetadata(diffOutput string) string {
+	const linesToCheck = 2
+	splits := strings.SplitN(diffOutput, "\n", linesToCheck+1)
+	out := make([]string, 0, 1)
+	linesToIgnorePrefixes := []string{
+		"diff --git",
+		"index ",
```

Comment:
Is the extra space needed? Seems the below code still work without the extra. ` `

Reply:
It's not needed, but it is helpful. These lines look like `index foo..bar` and we want to match the largest possible prefix to try to ensure we're matching the right line.